src/common/libzmqutil/zap.c

```diff
@@ -0,0 +1,222 @@
+    if (!zf || zmsg_addmem (dst, zframe_data (zf), zframe_size (zf)) < 0)
+        return false;
+    return true;
+}
+
+static void logger (struct zmqutil_zap *zap,
+                    int severity,
+                    const char *fmt, ...)
+{
+    if (zap->logger) {
+        va_list ap;
+        char buf[256];
+
+        va_start (ap, fmt);
+        vsnprintf (buf, sizeof (buf), fmt, ap);
+        va_end (ap);
+
+        zap->logger (severity, buf, zap->logger_arg);
+    }
+}
+
+/* ZAP 1.0 messages have the following parts
+ * REQUEST                              RESPONSE
+ *   0: version                           0: version
+ *   1: sequence                          1: sequence
+ *   2: domain                            2: status_code
+ *   3: address                           3: status_text
+ *   4: identity                          4: user_id
+ *   5: mechanism                         5: metadata
+ *   6: client_key
+ */
+static void zap_cb (flux_reactor_t *r,
+                    flux_watcher_t *w,
+                    int revents,
+                    void *arg)
+{
+    struct zmqutil_zap *zap = arg;
+    zmsg_t *req = NULL;
+    zmsg_t *rep = NULL;
+    char pubkey[41];
+    const char *status_code = "400";
+    const char *status_text = "No access";
+    const char *user_id = "";
+    zcert_t *cert;
+    const char *name = NULL;
+    int log_level = LOG_ERR;
+
+    if ((req = zmsg_recv (zap->sock))) {
+        if (!streq_zmsg_nth (req, 0, "1.0")
+                || !streq_zmsg_nth (req, 5, "CURVE")
+                || !pubkey_zmsg_nth (req, 6, pubkey)) {
+            logger (zap, LOG_ERR, "ZAP request decode error");
+            goto done;
+        }
+        if ((cert = zcertstore_lookup (zap->certstore, pubkey)) != NULL) {
+            status_code = "200";
+            status_text = "OK";
+            user_id = pubkey;
+            name = zcert_meta (cert, "name");
+            log_level = LOG_INFO;
+        }
+        if (!name)
+            name = "unknown";
+        logger (zap, log_level, "overlay auth %s %s", name, status_text);
+
+        if (!(rep = zmsg_new ()))
+            goto done;
+        if (!add_zmsg_nth (rep, req, 0)
+                || !add_zmsg_nth (rep, req, 1)
+                || zmsg_addstr (rep, status_code) < 0
+                || zmsg_addstr (rep, status_text) < 0
+                || zmsg_addstr (rep, user_id) < 0
+                || zmsg_addmem (rep, NULL, 0) < 0) {
+            logger (zap, LOG_ERR, "ZAP response encode error");
+            goto done;
+        }
+        if (zmsg_send (&rep, zap->sock) < 0)
+            logger (zap, LOG_ERR, "ZAP send error");
+    }
+done:
+    zmsg_destroy (&req);
+    zmsg_destroy (&rep);
+}
+
+/* Create a zcert_t and add it to in-memory zcertstore_t.
+ */
+int zmqutil_zap_authorize (struct zmqutil_zap *zap,
+                           const char *name,
+                           const char *pubkey)
+{
+    uint8_t public_key[32];
+    zcert_t *cert;
+
```

Comment:
Since this code has been moved "external" to the broker, should all public functions here check for valid arguments? (.e. return EINVAL on `!zap || !name || !pubkey`)

I realize that will add extra development and testing time, so I'm fine with leaving that out for now, but thought I'd ask.

Reply:
I thought not.  Maybe we could add that and a local unit test if a second user emerges?  For now I wanted to keep moving forward.

Reply:
Should this file have just been kept under `src/broker` if there is no other user current or planned?

I guess it is moot at this point if we're just moving forward.

Reply:
It does seem like a useful standalone thing if we ever want to have a process other than the broker implement a zeromq-based server that uses 0MQ security, so I'd prefer we keep it here.  I think you've convinced me that I should take a small amount of time and make the interfaces more robust and add a unit test, so I'll go ahead and do that.